### PR TITLE
Refactor integer type inference logic to fit smallest type

### LINKED ISSUE #802: Improve CSV `infer_field_schema` to be more restrictive
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

`infer_field_schema` infers any integer to be an int64: https://github.com/apache/arrow-rs/blob/9ca59b6cc0571f86862836788f3a1ccf4e91f370/arrow/src/csv/reader.rs#L86. This means that data parsed as CSV with the automatic schema is going to be much larger than necessary. 

**Describe the solution you'd like**

Integers should only be as long as necessary to represent the numbers. Similarly, floats should only be as large as needed to accurately represent the numbers in the data. 

**Describe alternatives you've considered**

The current behavior works but doesn't use the format efficiently.

### PR DESCRIPTION
**Pull request by jondo2010**
# Which issue does this PR close?

Partially fixes #802

# Rationale for this change

As explained in #802, the current `infer_schema` logic always returns `DataType::Int64` for all integer types.

This PR adds additional inferring logic to find the smallest integer type that fits the data.

This PR also fixes an issue where the data in fact doesn't fit in a `DataType::Int64` and instead requires a `UInt64`. 

# Are there any user-facing changes?

Any user code loading CSV data using `infer_schema()` and always expecting to receive `Int64` fields would be affected.

No API changes.

### CODE REVIEW

**Review by Jefffrey — CHANGES_REQUESTED** [review]
Will need to swap out usage of lexical_core

I'll take a closer look at the PR later, just noting this down first :+1:

File: arrow-csv/src/reader/mod.rs [hunk 1]
```diff
@@ -157,6 +158,61 @@ lazy_static! {
...
+                } else {
+                    DataType::Int64
+                }
+            }
+            _ => DataType::Int64,
+        }
+    } else {
+        match s.len() {
+            1..=3 => {
+                if lexical_core::parse::<u8>(s.as_bytes()).is_ok() {
```
[the remark this answers is not in the document]
Comment (jondo2010) [reply arrow-csv/src/reader/mod.rs:191]: Ah that's unfortunate.

**Review by tustvold — COMMENTED** [review]
I worry that this will make schema inference very fragile, as it is typically only run on a subset of the data and therefore with this change may be overly restrictive on size or sign.

A few thoughts:

* We should prefer signed types by default, as this most of the time is sufficient
* The new behaviour should be opt-in
* I think this will substantially regress performance as it now must parse integers

Perhaps we could see how other systems handle this, e.g. DuckDb or Spark?

**Comment by tustvold** [comment]
Marking as draft as not waiting on review, please feel to mark as ready for review when you would like me to take another look

**Pull request closed without merging**